### ISSUE TITLE
stmgr did not quit by LOG(FATAL)

### ISSUE THREAD

**Issue by huijunw** [apache/incubator-heron#2482]
We observed stmgr stuck in a particular scenario.

In the heron-executor log, the stmgr restarted 3 times, but in the last restart, stmgr seems stuck and did not quit.

Here is the log:

```
[2017-10-22 10:13:05 +0000] [INFO]: stmgr-103 stdout: 2017-10-22 10:13:05,039:30074(0x7f9378667700):ZOO_ERROR@handle_socket_error_msg@1643: .....
[2017-10-22 10:13:05 +0000] [INFO]: stmgr-103 stdout: 2017-10-22 10:13:05,039:30074(0x7f9378667700):ZOO_INFO@check_events@1703: initiated connection to server [....]
[2017-10-22 10:13:05 +0000] [INFO]: stmgr-103 stdout: 2017-10-22 10:13:05,040:30074(0x7f9378667700):ZOO_INFO@check_events@1750: session establishment complete on server [...]...
[2017-10-22 10:13:15 +0000] [INFO]: stmgr-103 stdout: F1022 10:13:15.199504 30074 client.cpp:184] UnPackString failed from connection 0x2374f40 from hostport ....
[2017-10-22 10:13:15 +0000] [INFO]: stmgr-103 stdout: *** Check failure stack trace: ***
[2017-10-22 10:13:15 +0000] [INFO]: stmgr-103 stdout:     @           0x4d8cba  google::LogMessage::Fail()
[2017-10-22 10:13:15 +0000] [INFO]: stmgr-103 stdout:     @           0x4dab22  google::LogMessage::SendToLog()
[2017-10-22 10:13:15 +0000] [INFO]: stmgr-103 stdout:     @           0x4d889f  google::LogMessage::Flush()
[2017-10-22 10:13:15 +0000] [INFO]: stmgr-103 stdout:     @           0x4db43e  google::LogMessageFatal::~LogMessageFatal()
[2017-10-22 10:13:15 +0000] [INFO]: stmgr-103 stdout:     @           0x4b9902  Client::OnNewPacket()
[2017-10-22 10:13:15 +0000] [INFO]: stmgr-103 stdout:     @           0x4bbf39  Connection::handleDataRead()
[2017-10-22 10:13:15 +0000] [INFO]: stmgr-103 stdout:     @           0x4b7adc  BaseConnection::handleRead()
[2017-10-22 10:13:15 +0000] [INFO]: stmgr-103 stdout:     @           0x4bdccd  EventLoopImpl::handleReadCallback()
[2017-10-22 10:13:15 +0000] [INFO]: stmgr-103 stdout:     @           0x4c8a8a  event_process_active_single_queue
[2017-10-22 10:13:15 +0000] [INFO]: stmgr-103 stdout:     @           0x4c9307  event_base_loop
[2017-10-22 10:13:15 +0000] [INFO]: stmgr-103 stdout:     @           0x40945f  main
[2017-10-22 10:13:15 +0000] [INFO]: stmgr-103 stdout:     @     0x7f9378cbcb35  __libc_start_main
[2017-10-22 10:13:15 +0000] [INFO]: stmgr-103 stdout:     @           0x40d89f  (unknown)
[2017-10-22 10:13:15 +0000] [INFO]: stmgr-103 stdout:     @              (nil)  (unknown)
[2017-10-22 10:13:15 +0000] [INFO]: stmgr-103 (pid=30074) exited with status 134. command=['./heron-core/bin/heron-stmgr', .., 'ckptmgr-103']
[2017-10-22 10:13:25 +0000] [INFO]: Running stmgr-103 process as ./heron-core/bin/heron-stmgr .... ckptmgr-103
[2017-10-22 10:13:25 +0000] [INFO]: Logging pid 31007 to file stmgr-103.pid
[2017-10-22 10:13:25 +0000] [INFO]: stmgr-103 stdout: WARNING: Logging before InitGoogleLogging() is written to STDERR
[2017-10-22 10:13:25 +0000] [INFO]: stmgr-103 stdout: I1022 10:13:25.323081 31007 yaml-file-reader.cpp:55] Config file ./heron-conf/heron_internals.yaml changed. reloading..
[2017-10-22 10:13:25 +0000] [INFO]: stmgr-103 stdout: 2017-10-22 10:13:25,329:31007(0x7f22bc45b780):ZOO_INFO@log_env@712: Client environment:zookeeper.version=zookeeper C client 3.4.5
[2017-10-22 10:13:25 +0000] [INFO]: stmgr-103 stdout: 2017-10-22 10:13:25,329:31007(0x7f22bc45b780):ZOO_INFO@log_env@716: Client environment:...
...
[2017-10-22 10:13:25 +0000] [INFO]: stmgr-103 stdout: 2017-10-22 10:13:25,330:31007(0x7f22bc45b780):ZOO_INFO@zookeeper_init@786: Initiating client connection, host=...
[2017-10-22 10:13:25 +0000] [INFO]: stmgr-103 stdout: 2017-10-22 10:13:25,332:31007(0x7f22bac11700):ZOO_INFO@check_events@1703: initiated connection to server [...
[2017-10-22 10:13:25 +0000] [INFO]: stmgr-103 stdout: 2017-10-22 10:13:25,332:31007(0x7f22bac11700):ZOO_INFO@check_events@1750: session establishment complete on server [....
[2017-10-22 10:13:31 +0000] [INFO]: stmgr-103 stdout: F1022 10:13:31.112793 31007 client.cpp:184] UnPackString failed from connection 0x216af40 from hostport ...
[2017-10-22 10:13:31 +0000] [INFO]: stmgr-103 stdout: *** Check failure stack trace: ***
[2017-10-22 10:13:31 +0000] [INFO]: stmgr-103 stdout:     @           0x4d8cba  google::LogMessage::Fail()
[2017-10-22 10:13:31 +0000] [INFO]: stmgr-103 stdout:     @           0x4dab22  google::LogMessage::SendToLog()
[2017-10-22 10:13:31 +0000] [INFO]: stmgr-103 stdout:     @           0x4d889f  google::LogMessage::Flush()
[2017-10-22 10:13:31 +0000] [INFO]: stmgr-103 stdout:     @           0x4db43e  google::LogMessageFatal::~LogMessageFatal()
[2017-10-22 10:13:31 +0000] [INFO]: stmgr-103 stdout:     @           0x4b9902  Client::OnNewPacket()
[2017-10-22 10:13:31 +0000] [INFO]: stmgr-103 stdout:     @           0x4bbf39  Connection::handleDataRead()
[2017-10-22 10:13:31 +0000] [INFO]: stmgr-103 stdout:     @           0x4b7adc  BaseConnection::handleRead()
[2017-10-22 10:13:31 +0000] [INFO]: stmgr-103 stdout:     @           0x4bdccd  EventLoopImpl::handleReadCallback()
[2017-10-22 10:13:31 +0000] [INFO]: stmgr-103 stdout:     @           0x4c8a8a  event_process_active_single_queue
[2017-10-22 10:13:31 +0000] [INFO]: stmgr-103 stdout:     @           0x4c9307  event_base_loop
[2017-10-22 10:13:31 +0000] [INFO]: stmgr-103 stdout:     @           0x40945f  main
[2017-10-22 10:13:31 +0000] [INFO]: stmgr-103 stdout:     @     0x7f22bb266b35  __libc_start_main
[2017-10-22 10:13:31 +0000] [INFO]: stmgr-103 stdout:     @           0x40d89f  (unknown)
[2017-10-22 10:13:31 +0000] [INFO]: stmgr-103 stdout:     @              (nil)  (unknown)
[2017-10-22 10:13:31 +0000] [INFO]: stmgr-103 (pid=31007) exited with status 134. command=['./heron-core/bin/heron-stmgr', '....
[2017-10-22 10:13:41 +0000] [INFO]: Running stmgr-103 process as ./heron-core/bin/heron-stmgr ....
[2017-10-22 10:13:41 +0000] [INFO]: Logging pid 31175 to file stmgr-103.pid
[2017-10-22 10:13:41 +0000] [INFO]: stmgr-103 stdout: WARNING: Logging before InitGoogleLogging() is written to STDERR
[2017-10-22 10:13:41 +0000] [INFO]: stmgr-103 stdout: I1022 10:13:41.513188 31175 yaml-file-reader.cpp:55] Config file ./heron-conf/heron_internals.yaml changed. reloading..
[2017-10-22 10:13:41 +0000] [INFO]: stmgr-103 stdout: 2017-10-22 10:13:41,518:31175(0x7f4f45d6d780):ZOO_INFO@log_env@712: Client environment:zookeeper.version=zookeeper C client 3.4.5
....
[2017-10-22 10:13:41 +0000] [INFO]: stmgr-103 stdout: 2017-10-22 10:13:41,522:31175(0x7f4f45d6d780):ZOO_INFO@zookeeper_init@786: Initiating client connection, ....
[2017-10-22 10:14:01 +0000] [INFO]: stmgr-103 stdout: 2017-10-22 10:14:01,535:31175(0x7f4f44523700):ZOO_ERROR@handle_socket_error_msg@1643: Socket [....
[2017-10-22 10:14:01 +0000] [INFO]: stmgr-103 stdout: 2017-10-22 10:14:01,535:31175(0x7f4f44523700):ZOO_INFO@check_events@1703: initiated connection to server [....
[2017-10-22 10:14:01 +0000] [INFO]: stmgr-103 stdout: 2017-10-22 10:14:01,536:31175(0x7f4f44523700):ZOO_INFO@check_events@1750: session establishment complete on server [...
[2017-10-22 10:14:11 +0000] [INFO]: stmgr-103 stdout: F1022 10:14:11.689586 31175 client.cpp:184] UnPackString failed from connection 0x3bf2f40 from hostport ....
[2017-10-22 10:14:11 +0000] [INFO]: stmgr-103 stdout: *** Check failure stack trace: ***
[2017-10-22 10:14:11 +0000] [INFO]: stmgr-103 stdout:     @           0x4d8cba  google::LogMessage::Fail()
[2017-10-22 10:14:11 +0000] [INFO]: stmgr-103 stdout:     @           0x4dab22  google::LogMessage::SendToLog()
[2017-10-22 10:14:11 +0000] [INFO]: stmgr-103 stdout:     @           0x4d889f  google::LogMessage::Flush()
[2017-10-22 10:14:11 +0000] [INFO]: stmgr-103 stdout:     @           0x4db43e  google::LogMessageFatal::~LogMessageFatal()
[2017-10-22 10:14:11 +0000] [INFO]: stmgr-103 stdout:     @           0x4b9902  Client::OnNewPacket()
[2017-10-22 10:14:11 +0000] [INFO]: stmgr-103 stdout:     @           0x4bbf39  Connection::handleDataRead()
[2017-10-22 10:14:11 +0000] [INFO]: stmgr-103 stdout:     @           0x4b7adc  BaseConnection::handleRead()
[2017-10-22 10:14:11 +0000] [INFO]: stmgr-103 stdout:     @           0x4bdccd  EventLoopImpl::handleReadCallback()
[2017-10-22 10:14:11 +0000] [INFO]: stmgr-103 stdout:     @           0x4c8a8a  event_process_active_single_queue
[2017-10-22 10:14:11 +0000] [INFO]: stmgr-103 stdout:     @           0x4c9307  event_base_loop
[2017-10-22 10:14:11 +0000] [INFO]: stmgr-103 stdout:     @           0x40945f  main
[2017-10-22 10:14:11 +0000] [INFO]: stmgr-103 stdout:     @     0x7f4f44b78b35  __libc_start_main
[2017-10-22 10:14:11 +0000] [INFO]: stmgr-103 stdout:     @           0x40d89f  (unknown)
[2017-10-22 10:14:11 +0000] [INFO]: stmgr-103 stdout:     @              (nil)  (unknown)
[2017-10-23 10:33:07 +0000] [WARNING]: Connection dropped: outstanding heartbeat ping not received
[2017-10-23 10:33:07 +0000] [WARNING]: Transition to CONNECTING
[2017-10-23 10:33:07 +0000] [INFO]: Zookeeper connection lost
[2017-10-23 10:33:07 +0000] [INFO]: Connection state changed to: SUSPENDED
[2017-10-23 10:33:07 +0000] [INFO]: Connecting to ....

```

I expected `exited with status 134.` at 10:14:11, but there was not. Thus I think that stmgr stuck at  OG(FATAL).

**Comment by huijunw** [comment]
any thoughts to fix this @srkukarni

**Comment by srkukarni** [comment]
LOG(FATAL) is supposed to kill the process.
One thing that we could do is maybe add an explicit ::exit() after FATAL?

**Comment by objmagic** [comment]
This rarely happens so I'd like to keep the current state and wait if we see more of this error.

**Comment by nwangtw** [comment]
I believe LOG(FATAL) calls exit() to quit. If the process freezes, it seems the code is not executed. Calling ::exit() doesn't help.

The root cause is to solve the freezing issue itself.

**Comment by huijunwu** [comment]
should be fixed in #2908, cannot reproduce in the recent releases
